### ISSUE TITLE
Should have option to use output weights when validating

### ISSUE THREAD

**Issue by wxs** [fchollet/keras#496]
We discussed this in passing before, but I think we need to discuss it directly. There are at least some applications where you really do need to use `sample_weight` in validation as well as in training. This should at least be an option (I'll be building this for my uses either way, but would like to do so in such a way that I can merge it in to the master branch)

One use of sample_weight is to mask output values when you're doing sequence to sequence learning with different length sequences. If you don't also use sample_weight in validation, your loss will be hugely increased by all the meaningless masked output values.

See, for instance, the discussion happening over in #451 

This applies to `class_weight` as well when you are classifying inputs and have very different numbers of training examples for each class, but want to remove that bias in your training. If I have 99 examples in class A for each example in class B I'd like to be able to, rather than throwing away 98% of my training examples to normalize class counts, rather pass in a class_weight of 1/99 to class A, and get equivalent validation loss.

**Comment by fchollet** [comment]
Right, I think it makes sense to have it as an option.

**Comment by cmyr** [comment]
I've been investigating this as well, and I'm a bit confused. As far as I can tell, if you're using the `validation_split` argument of `fit()`, then output weights _are_ applied for validation as well as training, because `X`, `y`, and `sample_weight` are all standardized before being passed along to `_fit`; in _this_ case, it seems like there's no option to turn weights _off_ for validation, although I might be misreading something?

Anyway, my current use-case requires using `sample_weight` in validation, and also passing in distinct validation data. I've added weights to validation by adding a `weight_validation` flag and allowing `validation_data` to be a 3-tuple; and since the `class_weight` and `sample_weight` arguments seem mutually exclusive, the 3rd item in the tuple is presumed to be whichever of those arguments is present.

It's ugly, basically.

Do you guys have any thoughts on this? I'm concerned there are intricacies to the model code that I don't fully appreciate, and this weights stuff feels a bit fragile in general.

**Comment by fchollet** [comment]
> this weights stuff feels a bit fragile in general.

Agreed... we should probably rewrite from scratch.

**Comment by wxs** [comment]
I put some thoughts over in #600 about one aspect of a rewrite: I think output_mask should be a separate parameter from sample_weight because of normalization concerns.

w.r.t. what the best API is for providing all these parameters in the case of auto-validation split vs. explicit validation split, I think that's another. I agree that passing in a 4-tuple `(X, y, sample_weight, output_mask)` sounds pretty horrible..

One proposal: I could imagine a Dataset type that included X, y, sample_weight, class_weight, and output_mask, and that you would pass in a Dataset to fit, and optionally a validation Dataset.

**Comment by fchollet** [comment]
I am adding the option to use validation sample weights. They would simply be passed optionally as part of the `validation_data` tuple.

Additionally I have noticed that in the case where `validation_split` is used, `class_weight` is applied to the validation data, which is not the expected behavior (classes are of course supposed unknown for the validation data). I am also fixing this.

**Comment by cmyr** [comment]
@fchollet excellent to hear, looking forward to testing this.

**Comment by fchollet** [comment]
@cmyr the fix is in. Let me know if that works for you.

**Comment by cmyr** [comment]
Great I'll test with this tomorrow morning.

**Comment by fchollet** [comment]
I reverted the fix, as it seemed to create other issues (although it did pass all the tests...). Now investigating what went wrong, and writing tests for it.

**Comment by fchollet** [comment]
@cmyr the fix is back up. Let me know!

**Comment by cmyr** [comment]
cool, this was very similar to the implementation I had, which is encouraging. Have some trainings running, will let you know if anything seems awry.

**Comment by dc-ai** [comment]
It looks like class_weight still does not get factored into validation loss calculations?

And it seems that fit_generator does not support having either class or sample weights being used when calculating validation loss?

**Comment by skarakulak** [comment]
It would be nice to have a class weighted validation loss without the need to modify the loss function itself.
